### PR TITLE
Improvements to the merge script.

### LINKED ISSUE #7098: [dev tools] merge-pr.sh please provide a way to override REMOTE_URL
Some of us may have a different `origin` branch as a safety measure, so it'd be useful to have:
```
merge-pr.sh --remote=official-gh-write-permissions
```

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure.

Fixes / closes #7098

See commit log for details.

### CODE REVIEW

**Comment by ejgallego** [comment]
I get the same result, but thanks anyways. I think I am just bothering you folks too much, I'll live with the warning as I know what I'm doing here.

**Comment by ejgallego** [comment]
if you are curious the warning I get is:
```
error: remote origin does not point to the official Coq repo 
error: that is git@github.com:coq/coq 
error: it points to https://github.com/coq/coq.git instead 
```

**Comment by SkySkimmer** [comment]
So we need to port the following to .sh?
https://github.com/raxod502/straight.el/blob/0782e761ff86dea080a6873d0429192beaa6e44b/straight.el#L1439-L1445

**Comment by Zimmi48** [comment]
OK so I misunderstood what you meant by "read-only". Anyways, we can fix this by adding `https://github.com/coq/coq.git` and `https://github.com/coq/coq` to the list of remotes that do not generate a warning. We already explicitly list `git@github.com:coq/coq.git` and `git@github.com:coq/coq` so I don't see the problem in adding two more.

**Comment by ejgallego** [comment]
Honestly I think I misunderstood the importance of the warning and I am bikeshedding here and taking time from you folks, sorry about that. From my part feel free to close the bug, or just add an equivalent URL list if you feel like it.

**Comment by Zimmi48** [comment]
Done. Please test :)

**Comment by ejgallego** [comment]
It works thanks!

File: dev/tools/merge-pr.sh [hunk 1]
```diff
@@ -77,11 +78,13 @@ if [ -z "$REMOTE" ]; then
   error "please run: git branch --set-upstream-to=THE_REMOTE/$CURRENT_LOCAL_BRANCH"
   exit 1
 fi
-REMOTE_URL=$(git remote get-url "$REMOTE" --push)
-if [ "$REMOTE_URL" != "$OFFICIAL_REMOTE_URL" -a \
-     "$REMOTE_URL" != "$OFFICIAL_REMOTE_URL.git" ]; then
+REMOTE_URL=$(git remote get-url "$REMOTE" --all)
+if [ "$REMOTE_URL" != "$OFFICIAL_REMOTE_GIT_URL" -a \
+     "$REMOTE_URL" != "$OFFICIAL_REMOTE_GIT_URL.git" -a \
+     "$REMOTE_URL" != "$OFFICIAL_REMOTE_HTTPS_URL" -a \
```
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: Please don't use `-a` as bash can parse it in unintended ways.

Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: OK. Anyways, it is going to conflict with the shell script PR. Is it ready to be merged? If yes, I'll rebase afterwards.

**Review by maximedenes — DISMISSED** [review]

**Comment by maximedenes** [comment]
Waiting for #6809 to be merged first.

**Comment by Zimmi48** [comment]
I rebased and took the opportunity to add two more commits to this PR, following @herbelin's comments https://github.com/coq/coq/issues/7097#issuecomment-378632415 and https://github.com/coq/coq/issues/7097#issuecomment-379015847 (and also @maximedenes's comment https://github.com/coq/coq/pull/7074#pullrequestreview-107049681 and my answer there).

This last commit is:

> Add sanity check in merge script: local branch is up-to-date.
>
> This check will fail whenever several PRs are merged in a row without pushing, but then the merger is supposed to understand what's happening and why it's okay to bypass the check.

I think it's a good check to have and its limitation is not problematic (not many PRs are merged in a row with the new distributed merging model). But if people feel otherwise, I can drop this commit or move it to a new PR...

**Comment by SkySkimmer** [comment]
We could use `git merge-base --is-ancestor` to print a different message when the difference with upstream is only that we have more commits (this is the case when merging multiple PRs without pushing between them, but also if we foolishly committed on master).

File: dev/tools/merge-pr.sh [hunk 2]
```diff
@@ -79,7 +79,7 @@ if [ -z "$REMOTE" ]; then
   error "please run: git branch --set-upstream-to=THE_REMOTE/$CURRENT_LOCAL_BRANCH"
   exit 1
 fi
-REMOTE_URL=$(git remote get-url "$REMOTE" --push)
+REMOTE_URL=$(git remote get-url "$REMOTE" --all)
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: Indeed, it is not true but I agree it is confusing.

My `.git/config` contains:

```
[remote "upstream"]
	url = git@github.com:coq/coq
	fetch = +refs/heads/*:refs/remotes/upstream/*
```

`git remote get-url upstream --push` and `git remote get-url --all upstream` both give `git@github.com:coq/coq`

If I do `git remote set-url --push upstream https://github.com/coq/coq` it then contains:

```
[remote "upstream"]
	url = git@github.com:coq/coq
	fetch = +refs/heads/*:refs/remotes/upstream/*
	pushurl = https://github.com/coq/coq
```

`git remote get-url upstream --push` now gives `https://github.com/coq/coq` and `git remote get-url --all upstream` gives `git@github.com:coq/coq`

[the remark this answers is not in the document]
Comment (SkySkimmer) [reply dev/tools/merge-pr.sh]: Actually it's possible to have multiple urls:
```bash
$ git remote set-url --add upstream foo
$ git remote get-url --all upstream
git@github.com:coq/coq.git
foo
```

**Comment by Zimmi48** [comment]
> We could use `git merge-base --is-ancestor` to print a different message when the difference with upstream is only that we have more commits (this is the case when merging multiple PRs without pushing between them, but also if we foolishly committed on master).

So for instance, print an error in case it is not an ancestor but just a warning if it is?
I updated the PR to do just that.

File: dev/tools/merge-pr.sh [hunk 3]
```diff
@@ -107,6 +110,25 @@ if [ "$BASE_BRANCH" != "coq:$CURRENT_LOCAL_BRANCH" ]; then
...
+LOCAL_BRANCH_COMMIT=$(git rev-parse HEAD)
+UPSTREAM_COMMIT=$(git rev-parse @{u})
+if [ "$LOCAL_BRANCH_COMMIT" != "$UPSTREAM_COMMIT" ]; then
+
+    # Is it just that the upstream branch is behind?
+    # It could just be that we merged other PRs and we didn't push yet
+
+    if git merge-base --is-ancestor -- "$UPSTREAM_COMMIT" "$LOCAL_BRANCH_COMMIT"; then
+        warning "Your branch is ahead of ${REMOTE}."
+        warning "You should see this warning only if you just merged another PR and did not push yet."
```
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: I would still ask confirmation

File: dev/tools/merge-pr.sh [hunk 4]
```diff
@@ -107,6 +110,25 @@ if [ "$BASE_BRANCH" != "coq:$CURRENT_LOCAL_BRANCH" ]; then
...
+# Sanity check: the local branch is up-to-date with upstream
+
+LOCAL_BRANCH_COMMIT=$(git rev-parse HEAD)
+UPSTREAM_COMMIT=$(git rev-parse @{u})
+if [ "$LOCAL_BRANCH_COMMIT" != "$UPSTREAM_COMMIT" ]; then
+
+    # Is it just that the upstream branch is behind?
+    # It could just be that we merged other PRs and we didn't push yet
+
+    if git merge-base --is-ancestor -- "$UPSTREAM_COMMIT" "$LOCAL_BRANCH_COMMIT"; then
```
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: No `--`

File: dev/tools/merge-pr.sh [hunk 5]
```diff
@@ -107,6 +110,25 @@ if [ "$BASE_BRANCH" != "coq:$CURRENT_LOCAL_BRANCH" ]; then
...
+        warning "Your branch is ahead of ${REMOTE}."
+        warning "You should see this warning only if you just merged another PR and did not push yet."
+    else
+        error "Local branch is not up-to-date with ${REMOTE}."
+        error "Pull before merging."
+        ask_confirmation
```
Comment (SkySkimmer) [inline dev/tools/merge-pr.sh]: Here we should not ask for confirmation, we should unconditionally error. If we merged and pushed we could lose commits.

File: dev/tools/merge-pr.sh (hunk 4 above)
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: Without it, it doesn't seem to support commits.

File: dev/tools/merge-pr.sh (hunk 3 above)
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: OK

File: dev/tools/merge-pr.sh (hunk 5 above)
Comment (Zimmi48) [reply dev/tools/merge-pr.sh]: Well, you would need to force-push to do that. And force-pushing is actually disabled for everyone (including admins) on our branches. However, we can imagine a use-case (if someone uses this script to merge in a branch in their fork and they want to force-push...).

**Comment by Zimmi48** [comment]
Really? I quite like it that it doesn't wait for a new line.

**Comment by SkySkimmer** [comment]
>Really? I quite like it that it doesn't wait for a new line.

It freaks me out every time even when it already asked 10s ago.

**Comment by Zimmi48** [comment]
Here you go @SkySkimmer. I didn't change the default behavior but if many people ask for it, we could. In the meantime, you can setup an alias.

**Review by maximedenes — APPROVED** [review]
LGTM

**Pull request merged**